Allow fan placement in any order in air loop hvac dedicated outdoor air system

lgu1234 opened the pull request:
Pull request overview
---------------------
 - Fixes Phase 1 in #9066
 - The current allowed fan placement is draw and blow through only. This limits fan locations. The fix allow the fan is placed any order by removing error checks for fan placement
 - Fix Phase 2 
 - Add fan heat to the cooling coil capacity size when a fan is placed before the cooling coil.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [X ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - diff is expected in an example file as SmallOffice_CentralDOAS.idf. This file can be used as a test file for Phase 2
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ X] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - 
[9066.idf.txt](https://github.com/NREL/EnergyPlus/files/9114922/9066.idf.txt)

 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

lgu1234 committed:
Upload a fix

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -543,14 +535,6 @@ namespace AirLoopHVACDOAS {
-                                                   CompName));
-                            ShowContinueError(state, format("The current position is number {}", CompNum));
-                            errorsFound = true;
-                        }
```
Remove error checks for fan placement, so that a fan can be placed in any order.

lgu1234 committed:
Add fan heat to cooling coil capacity size

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -522,14 +524,7 @@ namespace AirLoopHVACDOAS {
-                            ShowContinueError(state, format("The current position is number {}", CompNum));
-                            errorsFound = true;
-                        }
+                        FanOrder = CompNum; 
```
Assign CompNum to a fan component

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -543,14 +538,7 @@ namespace AirLoopHVACDOAS {
-                            ShowContinueError(state, format("The current position is number {}", CompNum));
-                            errorsFound = true;
-                        }
+                        FanOrder = CompNum;
```
Assign CompNum to a fan component

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -565,6 +553,7 @@ namespace AirLoopHVACDOAS {
                         if (errorsFound) { // is this really needed here, program fatals out later on when errorsFound = true
                             ShowFatalError(state, "GetAirLoopDOASInput: Program terminated for previous conditions.");
                         }
+                        CoolingCoilOrder = CompNum;
```
Assign CompNum to a cooling coil component

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -633,6 +622,7 @@ namespace AirLoopHVACDOAS {
                         thisOutletNodeNum = state.dataAirLoop->OutsideAirSys(thisDOAS.m_OASystemNum)
                                                 .compPointer[CompNum]
                                                 ->getAirOutNode(state, CompName, 0, OutletNodeErrFlag);
+                        CoolingCoilOrder = CompNum;
```
Assign CompNum to a cooling coil component

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -699,6 +689,9 @@ namespace AirLoopHVACDOAS {
                     }
+                    if (CoolingCoilOrder > FanOrder) {
+                        thisDOAS.FanHeatAdded = true;
+                    }
```
When a fan is placed before a cooling coil, FanHeat should be added in cooling coil capacity size

rraustad replied:
You don't even need FanBeforeCoolingCoilFlag do you. If CoolingCoilOrder > FanOrder then it is blow through.

rraustad replied:
What we have no is a single enum for fan placement.

    enum class FanPlacement
    {
        Invalid = -1,
        BlowThru,
        DrawThru,
        Num
    };

This fan location applies to both the cooling and heating coil. But the fan could actually be in between the coils. That can not happen now but could in the future. It would be better to have enums for BlowThruCooling and BlowThruHeating, and same for DrawThru, but we just would not use the heating enums yet. Which means we would not define the heating enums yet. Could you change FanBeforeCoolingCoilFlag to be DataAirSystems::FanPlacement::supFanLocation and use existing enums? That way we can expand on this in the future.

lgu1234 replied:
@rraustad Phase 1 allows a supply fan can be in any order. In other words, BlowThru and DrawThru restrictions are removed in this module. Phase 2 is implemented based on Phase 1. Therefore, it is not appropriate to use existing FanPlacement enum. Phase 2 checks locations of both fan and cooling coil. If a fan is placed before a cooling coil, the fan heat will be added in cooling capacity sizing.

lgu1234 commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -961,6 +954,31 @@ namespace AirLoopHVACDOAS {
+            Real64 CpAir = Psychrometrics::PsyCpAirFnW(state.dataLoopNodes->Node(this->m_FanInletNodeNum).HumRat);
+            Real64 fanDeltaT = designHeatGain / (sizingMassFlow * CpAir);
+            this->PrecoolTemp -= fanDeltaT;
+        }
```
Add fan heat

rraustad replied:
I was not expecting this. To size a coil with fan heat all that is needed is to set the fan index, type and placement. Then the call to CoolingCapacitySizing will include the fan heat (if that section of code is corrected) in coil sizing. Was this method investigated? Is there a reason that DOAS systems need to account for fan heat instead of using the sizing routine? The DOAS design air flow should be known when the DOAS coils are sized. If this section of code is filled in then the coils should size correctly. If not then I would like to know why.

    if (this->curOASysNum > 0) { // coil is in the OA stream
        // need to find fan type in OA system
    } else {
        switch (this->primaryAirSystem(this->curSysNum).supFanModelType) {

Myoldmopar replied:
@lgu1234 please provide a reply here and we'll work on pushing this along the review train.

lgu1234 replied:
@rraustad The mass flow is given. I tried to fill in the section code in CoolingCapacitySizing by calling CoolingCapacitySizer::size. You told me that before. For some reason, I was unable to connect the function size in AirLoopHVACDOAS. In other words, I was unable to set "auto coolingCoilSize &CoolingCapacitySizing". That was why I use the same functions in your this->calcFanDesHeatGain(DesVolFlow). Therefore, same functions are called to get fan heat. The difference is I set up locally.

rraustad replied:
You don't call CoolingCapacitySizer from AirLoopHVACDOAS, the cooling coil calls that function. What you do here is set OutsideAirSys().supFanModelType and OutsideAirSys().supFanLocation and let the sizing function calculate the fan heat gain, delta T, and adjustment to coil capacity. You are adding this for DOAS systems but there can also be coils in the OutsideAirSystem so be generic to allow coils in OA systems to also use this code. Then all you need to do is fill in "// need to find fan type in OA system" with similar code to what is below here.

    if (this->curOASysNum > 0) { // coil is in the OA stream
        // need to find fan type in OA system
        } else {
            switch (this->primaryAirSystem(this->curSysNum).supFanModelType) {
            case DataAirSystems::StructArrayLegacyFanModels: {
                FanCoolLoad = this->calcFanDesHeatGain(DesVolFlow);
                break;
            }
            case DataAirSystems::ObjectVectorOOFanSystemModel: {
                FanCoolLoad = this->calcFanDesHeatGain(DesVolFlow);
                break;
            }
            default:
                // do nothing
                break;
            } // end switch

Then you should not have to adjust the DeltaT here:

                } else if (this->curOASysNum > 0 && this->outsideAirSys(this->curOASysNum).AirLoopDOASNum > -1) {
                    auto &thisAirloopDOAS = this->airloopDOAS[this->outsideAirSys(this->curOASysNum).AirLoopDOASNum];
                    if (thisAirloopDOAS.m_FanIndex > -1 && thisAirloopDOAS.FanBlowTroughFlag &&
                        thisAirloopDOAS.m_FanTypeNum == SimAirServingZones::CompType::Fan_System_Object) {
                        int FanIndex = thisAirloopDOAS.m_FanIndex;
                        Real64 DeltaT = state.dataHVACFan->fanObjs[FanIndex]->getFanDesignTemperatureRise(state);
                        CoilInTemp += DeltaT;
                    }

lgu1234 replied:
@rraustad I debug the code and found that it never reach your section. Instead, code go through Line 296.

rraustad replied:
@lgu1234 that's because of the private "else if" for AirLoopHVACDOAS. That section of code should be merged into the else section. All that's needed is the inlet and outlet conditions and DesVolFlow to be set properly. If you run a file that has an OutdoorAirSystem it will go through the else section.

lgu1234 replied:
@rraustad Actually, I found the code reach line 275 in CoolingCapacitySizing. Your section is not reached. Here is a section of code:

                    if (thisAirloopDOAS.m_FanIndex > -1 && thisAirloopDOAS.FanBlowTroughFlag &&
                        thisAirloopDOAS.m_FanTypeNum == SimAirServingZones::CompType::Fan_System_Object) {
                        int FanIndex = thisAirloopDOAS.m_FanIndex;
                        Real64 DeltaT = state.dataHVACFan->fanObjs[FanIndex]->getFanDesignTemperatureRise(state);
                        CoilInTemp += DeltaT;
                    }

I already added fan heat, if the fan is blowtrough. Therefore, I need to add one more scenario when a fan is placed before the cooling coil.

rraustad replied:
OK, go ahead and use that else if block to size a coil for AirLoopHVACDOAS. Too much would need to be done to avoid conflicts with FinalSysSizing in the else part of those blocks.

lgu1234 committed:
Clang format

lgu1234 committed:
Minor correction

rraustad reviewed (changes requested):
Add draw through fan heat calculation for AirLoopHVACDOAS coils.

lgu1234 committed:
Address comments

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented:
@Myoldmopar @rraustad I moved the cooling coil capacity section with fan heat into CoolingCapacitySizing. I also removed the heating coil capacity section with fan heat in HeatingCapacitySizing. As @rraustad told me, no fan heat should be added in heat coil capacity for conservative approach. Please take a look and let me know your concerns.

lgu1234 committed:
Clang format

lgu1234 committed:
More clang format

lgu1234 committed:
Revise unit test

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented:
@rraustad I revised the unit test a lot. The comparisons are cooling coil air flow rate and cooling capacity with fan heat. Please let me know what you think.

lgu1234 committed:
clang format

lgu1234 committed:
One more clang format

Myoldmopar commented:
The build warning is about [this line](https://github.com/NREL/EnergyPlus/blob/55986bb4dea701888f351ff60237e23a5113fc5c/src/EnergyPlus/Autosizing/CoolingCapacitySizing.cc#L291): 

```
this->dataFanEnumType = this->dataFanEnumType = DataAirSystems::StructArrayLegacyFanModels;
```

which is obviously a typo and should just set the enum type with one equals operation.

On the debug build, the CoolingCapacitySizingGauntlet is [failing](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-55986bb4dea701888f351ff60237e23a5113fc5c-x86_64-Linux-Ubuntu-18.04-gcc-7.5-UnitTestsCoverage-Debug.html) with an array index out of bounds error that should be relatively easy to solve.

Once those are fixed up, maybe this will be ready, but back in your court for now @lgu1234 .

rraustad commented on src/EnergyPlus/Autosizing/CoolingCapacitySizing.cc:
```diff
@@ -275,10 +275,30 @@ Real64 CoolingCapacitySizer::size(EnergyPlusData &state, Real64 _originalValue,
+                        Real64 CpAir = Psychrometrics::PsyCpAirFnW(state.dataLoopNodes->Node(thisAirloopDOAS.m_FanInletNodeNum).HumRat);
+                        DeltaT = FanCoolLoad / (thisAirloopDOAS.SizingMassFlow * CpAir);
                         CoilInTemp += DeltaT;
                     }
```
Where is the draw thru fan accounted for? For draw thru the coil outlet temp would need to be lowered by fan delta T.

lgu1234 replied:
@rraustad I added fan heat to the cooling coil, if a fan is placed before the cooling coil. Do we need to add heat heat to the cooling coil, if a fan is placed after the cooling coil?

rraustad replied:
The OA coil will be sized to meet the Precool temp which is at the outlet of the fan and coils. If a blow thru fan, fan heat is added to OA temp at the inlet of the coil. If draw thru, fan heat is subtracted from Precool temp and used as the coil outlet temp used for sizing so that after the fan the temperature will equal Precool temp.

    Sizing:System,
      11.0,                    !- Precool Design Temperature {C}

lgu1234 committed:
Fix a unit test failure and add one more section in the same unit test

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented:
@Myoldmopar A failure in CoolingCapacitySizingGauntlet unit test was fixed and the code was uploaded.
@rraustad I also add a new section in the CoolingCapacitySizingGauntlet unit test to provide a test for cooling capacity sizing without fan heat. The existing section tested cooling capacity sizing with fan heat.

The other comments will be addressed later.

lgu1234 committed:
Clang format

Myoldmopar commented:
> The other comments will be addressed later.

@lgu1234 thanks for the latest changes, but can you clarify what you mean here?  Are you saying you will be pushing more commits to this branch later?  Or you are saying this PR is complete and any other comments will be addressed at a later time.  I just want to make sure I don't spend time reviewing until the right time.  @rraustad you had requested changes, so at an appropriate time, could you see if your review comments have been addressed?

lgu1234 commented:
@rraustad The unit test of AirLoopHVACDOAS_TestOACompFanNoDrawAndBlow is to allow fan placed in any order. Nothing to do with cooling coil capacity. The unit test of AirLoopHVACDOAS_TestFanHeatAddeToCoolingCoilSize is to test cooling coil capacity with added fan heat.

In addition, I created a test case 22 to test cooling capacity without added fan heat in CoolingCapacitySizingGauntlet. The existing test case 20 in the same unit test tested cooling capacity size with the fan heat.

lgu1234 committed:
Adjust PreCoolTemp with drawthru fan heat

lgu1234 commented:
@rraustad The PrecoolTemp was adjusted with drawthrough fan heat. The code revision was uploaded.

Myoldmopar commented:
@rraustad could you take a look at the changes since your previous review and, if you are satisfied, mark it approved on your end?  I'll hold off any testing in case there are further changes to be made.

rraustad commented on src/EnergyPlus/AirLoopHVACDOAS.cc:
```diff
@@ -913,7 +907,7 @@ namespace AirLoopHVACDOAS {
+            if (!this->FanBeforeCoolingCoilFlag) {
                 state.dataLoopNodes->Node(state.dataAirLoop->OutsideAirSys(this->m_OASystemNum).InletNodeNum(1)).MassFlowRateMaxAvail =
                     this->SumMassFlowRate;
             }
```
Why does fan placement affect when the OA system MassFlowRateMaxAvail is set at the entrance of the OA system?

lgu1234 replied:
@rraustad This section is not necessary. I testes 3 cases with different fan placements: Draw, Blow and Middle. No differences were found. This section is removed.

rraustad commented:
This looks about how I expected. I had 1 question on why MaxAvail was written only if the fan was not the first component but other than that this looks good.

lgu1234 committed:
Address a comment

lgu1234 committed:
Merge branch 'develop' into Allow-fan-placement-in-any-order-in-AirLoopHVACDedicatedOutdoorAirSystem

lgu1234 commented:
@Myoldmopar @rraustad The last comment from Rich was addressed by removing the unnecessary section. It is ready for further review.

lgu1234 committed:
Correct a value for a unit test

Myoldmopar commented:
@rraustad there was one commit (7 days ago) since you had your max avail question (10 days ago).  Could you mark the PR approved if you feel like that one trailing piece is satisfied?  I will pull and test this locally now in anticipation of merging.

rraustad commented:
@Myoldmopar this looks good now and I approve.

Myoldmopar commented:
Thanks @rraustad , it's all passing locally too.  Merging this one.  Thanks @lgu1234
